### ISSUE TITLE
CPP: Add query for CWE-1041 Use of Redundant Code

### ISSUE THREAD

**Issue by ihsinme** [github/securitylab#443]
## Query

https://github.com/github/codeql/pull/6710

## CVE ID(s)

CVE-2021-29623

## Report

this query looks for functions for which there are wrapper functions with validation of arguments or work results.
the presence of wrapper functions already speaks of the developer's desire to ensure the launch of the function with additional safety. however, situations are possible when the developer has created a purely functional shell,
In this case, the detector will be useful in terms of assessing the need for an additional wrapper function.

I want to draw your attention to the presented representation of the problem, and not to the desire to fix it on the part of the developers (.

- [x] Are you planning to discuss this vulnerability submission publicly? (Blog Post, social networks, etc). *We would love to have you spread the word about the good work you are doing*

## Result(s)

- https://github.com/jackaudio/jack2/pull/796

**Comment by ihsinme** [comment]
good afternoon @JarLob.
Thank you for your prompt and detailed response.
he explains a lot to me.
1. Regarding the false detection in `PushItemWidth(80.0f);` and `DetailCtx(C);`. I propose to exclude them as follows.
1.1. exclude constants in base call
1.2. exclude global variables in wrapper function
1.3. exclude detection if there are multiple wrapper functions.

2.about your example from CVE-2021-29623. It seems to me that there is no return value validation that should rule out this false detection.

Thanks again for your constructive attitude towards improving detection. this is really important.

In order not to discuss several query options, I propose to make our communication linear. You are sending false detection kits, I am correcting the request.

if you agree with this, I will await the results of your analysis.
if not, I will wait for your suggestions.

thanks again.

**Comment by JarLob** [comment]
Sorry for the delay @ihsinme,
I think I have sent you enough false positives already. Below is the query I ran. Unfortunately the PR is closed already. You may open a new one or paste an updated version right here.
```ql
/**
 * @name Missed opportunity to call wrapper function
 * @description If a wrapper function is defined for a given function, any call to the given function should be via the wrapper function.
 * @kind problem
 * @id cpp/call-to-function-without-wrapper
 * @problem.severity warning
 * @precision medium
 * @tags correctness
 *       maintainability
 *       security
 *       external/cwe/cwe-1041
 */

import cpp
import semmle.code.cpp.valuenumbering.GlobalValueNumbering
import semmle.code.cpp.commons.Assertions

/**
 * A function call that is used in error situations (logging, throwing an exception, abnormal termination).
 */
class CallUsedToHandleErrors extends FunctionCall {
  CallUsedToHandleErrors() {
    // call that is known to not return
    not exists(this.(ControlFlowNode).getASuccessor())
    or
    // call throwing an exception
    exists(ThrowExpr tex | tex = this.(ControlFlowNode).getASuccessor())
    or
    // call logging a message, possibly an error
    exists(FormattingFunction ff | ff = this.(ControlFlowNode).getASuccessor())
    or
    // enabling recursive search
    exists(CallUsedToHandleErrors fr | getTarget() = fr.getEnclosingFunction())
  }
}

/** Holds if the conditions for a call outside the wrapper function are met. */
predicate conditionsOutsideWrapper(FunctionCall fcp) {
  fcp.getNumberOfArguments() > 0 and
  not exists(ConditionalStmt cdtmp | fcp.getEnclosingStmt().getParentStmt*() = cdtmp) and
  not exists(Loop lptmp | fcp.getEnclosingStmt().getParentStmt*() = lptmp) and
  not exists(ReturnStmt rttmp | fcp.getEnclosingStmt().getParentStmt*() = rttmp) and
  not exists(FunctionCall fctmp2 | fcp = fctmp2.getAnArgument().getAChild*()) and
  not exists(Assignment astmp | fcp = astmp.getRValue().getAChild*()) and
  not exists(Initializer intmp | fcp = intmp.getExpr().getAChild*()) and
  not exists(Assertion astmp | fcp = astmp.getAsserted().getAChild*()) and
  not exists(Operation optmp | fcp = optmp.getAChild*()) and
  not exists(ArrayExpr aetmp | fcp = aetmp.getAChild*()) and
  not exists(ExprCall ectmp | fcp = ectmp.getAnArgument().getAChild*())
}

/** Holds if the conditions for calling `fcp` inside the `fnp` wrapper function are met. */
pragma[inline]
predicate conditionsInsideWrapper(FunctionCall fcp, Function fnp) {
  not exists(FunctionCall fctmp2 |
    fctmp2.getEnclosingFunction() = fnp and fcp = fctmp2.getAnArgument().getAChild*()
  ) and
  not fcp instanceof CallUsedToHandleErrors and
  not fcp.getAnArgument().isConstant() and
  fcp.getEnclosingFunction() = fnp and
  fnp.getNumberOfParameters() > 0 and
  // the call arguments must be passed through the arguments of the wrapper function
  forall(int i | i in [0 .. fcp.getNumberOfArguments() - 1] |
    globalValueNumber(fcp.getArgument(i)) = globalValueNumber(fnp.getAParameter().getAnAccess())
  ) and
  // there should be no more than one required call inside the wrapper function
  not exists(FunctionCall fctmp |
    fctmp.getTarget() = fcp.getTarget() and
    fctmp.getFile() = fcp.getFile() and
    fctmp != fcp and
    fctmp.getEnclosingFunction() = fnp
  ) and
  // inside the wrapper function there should be no calls without paths to the desired function
  not exists(FunctionCall fctmp |
    fctmp.getEnclosingFunction() = fnp and
    fctmp.getFile() = fcp.getFile() and
    fctmp != fcp and
    (
      fctmp = fcp.getAPredecessor+()
      or
      not exists(FunctionCall fctmp1 |
        fctmp1 = fcp and
        (
          fctmp.getASuccessor+() = fctmp1 or
          fctmp.getAPredecessor+() = fctmp1
        )
      )
    )
  )
}

/** Holds if the conditions for the wrapper function are met. */
pragma[inline]
predicate conditionsForWrapper(FunctionCall fcp, Function fnp) {
  not exists(ExprCall ectmp | fnp = ectmp.getEnclosingFunction()) and
  not exists(Loop lp | lp.getEnclosingFunction() = fnp) and
  not exists(SwitchStmt sw | sw.getEnclosingFunction() = fnp) and
  not fnp instanceof Operator and
  // inside the wrapper function there should be checks of arguments or the result,
  // perhaps by means of passing the latter as an argument to some function
  (
    exists(IfStmt ifs |
      ifs.getEnclosingFunction() = fnp and
      (
        globalValueNumber(ifs.getCondition().getAChild*()) = globalValueNumber(fcp.getAnArgument()) and
        ifs.getASuccessor*() = fcp
        or
        ifs.getCondition().getAChild() = fcp
      )
    )
    or
    exists(FunctionCall fctmp |
      fctmp.getEnclosingFunction() = fnp and
      globalValueNumber(fctmp.getAnArgument().getAChild*()) = globalValueNumber(fcp)
    )
  ) and
  // inside the wrapper function there must be a function call to handle the error
  exists(CallUsedToHandleErrors fctmp |
    fctmp.getEnclosingFunction() = fnp and
    forall(int i | i in [0 .. fnp.getNumberOfParameters() - 1] |
      fnp.getParameter(i).getAnAccess().getTarget() =
        fcp.getAnArgument().(VariableAccess).getTarget() or
      fnp.getParameter(i).getUnspecifiedType() instanceof Class or
      fnp.getParameter(i).getUnspecifiedType().(ReferenceType).getBaseType() instanceof Class or
      fnp.getParameter(i).getAnAccess().getTarget() =
        fctmp.getAnArgument().(VariableAccess).getTarget()
    )
  )
}

from FunctionCall fc, Function fn
where
  exists(FunctionCall fctmp |
    conditionsInsideWrapper(fctmp, fn) and
    conditionsForWrapper(fctmp, fn) and
    conditionsOutsideWrapper(fc) and
    fctmp.getTarget() = fc.getTarget() and
    fc.getEnclosingFunction() != fn and
    fc.getEnclosingFunction().getMetrics().getNumberOfCalls() > fn.getMetrics().getNumberOfCalls()
  )
select fc, "Consider changing the call to $@", fn, fn.getName()
```

**Comment by ihsinme** [comment]
good day @JarLob 
thanks for the time you spent on this request.

> I think I have sent you enough false positives already.

Perhaps I am wrong, but I was expecting information from you about a false detection in this version of the request.

> >I reduce the width of the CallUsedToHandleErrors predicate, leaving only the exception detection in it. this results in decreased detection of https://lgtm.com/query/4138134136329408757/.

>I'll run the modified query on LGTM to see the results. Will get back to you.

If you find this request satisfactory, then I will try to amend the PR.

**Comment by JarLob** [comment]
> Perhaps I am wrong, but I was expecting information from you about a false detection in this version of the request.

Ah, sorry, it was hard to get recall all context. I did run the query on the whole LGTM back then after all. Even if I read a message "Consider changing the call to ***" as "You do not check parameters or return value of the function as you did in ***" (You didn't make the change to the message in query), the following finding do not make sense to me:
1. https://github.com/chemfiles/chemfiles `m_source( std::forward< T >( in_source ) )` suggests looking at `check_not_empty`
2. https://github.com/Tencent/phxrpc ` value_.Destroy(&::google::protobuf::internal::GetEmptyStringAlreadyInited(), GetArenaNoVirtual());` suggests looking at `SetAllocated`
3. https://github.com/soonho-tri/ibex-lib ` left[i] = p.first;` suggests looking at `var3BCID`
4. https://github.com/LIBOL/SOL `this->X_.resize(this->dim_);` suggests looking at `update_dim`

**Comment by ihsinme** [comment]
good afternoon @JarLob.
I made further changes to the request. (https://lgtm.com/query/713739777094140988/)
look at them on your kits, if they satisfy you and I will try to update PR

<details><summary>FIX https://github.com/chemfiles/chemfiles m_source( std::forward< T >( in_source ) ) suggests looking at check_not_empty</summary>
<p>

description in the next block

</p>
</details>

<details><summary>FIX https://github.com/Tencent/phxrpc value_.Destroy(&::google::protobuf::internal::GetEmptyStringAlreadyInited(), GetArenaNoVirtual()); suggests looking at SetAllocated</summary>
<p>

in this case i see two ways to limit false detection
1. restrict the use of a global variable inside the shell. this will exclude the wrappers created to work with these global objects.
2. restrict the ability to call functions using arguments that are function calls.

I settled on the second option (it seemed to me that it covers the problem more broadly)
so I added `not exists(FunctionCall fctmp2 | fctmp2 = fcp.getAnArgument().getAChild*()) and` to predicate `conditionsOutsideWrapper`.

this also removes false detection in https://github.com/chemfiles/chemfiles
</p>
</details>



<details><summary>FIX https://github.com/soonho-tri/ibex-lib left[i] = p.first; suggests looking at var3BCID</summary>
<p>

I decided to exclude any detection using the `Operator` call.
for this I inserted the constraint `not fcp.getTarget() instanceof Operator` into predicate `conditionsOutsideWrapper`

It is worth noting that a similar restriction exists for the wrapper function itself in predicate `conditionsForWrapper`. But it turned out to be insufficient(
</p>
</details>

<details><summary>FIX https://github.com/LIBOL/SOL this->X_.resize(this->dim_); suggests looking at update_dim</summary>
<p>

this example is tricky for me. 
the only criterion I would use to clean it up is if it has multiple wrappers. 
this will eliminate the situation when it is no longer a protective shell, but a functional shell.
but again this example was difficult for me, I am not even quite sure if this is a false detection.

in this case I introduced a restriction that the wrapper is not a virtual function. adding `not fnp instanceof VirtualFunction and` to predicate `conditionsInsideWrapper`
</p>
</details>

**Comment by JarLob** [comment]
We are getting less results, but IMHO they are not getting any better. Few from the top:
1. https://github.com/widelands/widelands `g_style_manager->font_style(tooltip_header_style_).as_font_tag(_("Game Mode"));` `Consider changing the call to as_listitem`.
2. https://github.com/pisa-engine/pisa explicit CapturedStream(int fd) : fd_(fd), uncaptured_fd_(dup(fd)) {` `Consider changing the call to dup`
3. https://github.com/janisozaur/openrct2-nsf `WindowInitScrollWidgets(this);` `Consider changing the call to window_ride_measurements_invalidate`

**Comment by ihsinme** [comment]
@JarLob 
I understand that your patience is running out, but I ask you not to lose hope and continue to give examples of FP.
Trust me this will make a good request.

my corrected version
https://lgtm.com/query/6911046000267266736/

<details><summary>my analysis of the FPs you found</summary>
<p>

1. example https://github.com/pisa-engine/pisa I believe is correct.
however, rather rough restrictions for other examples will rule out this case.(
2. example https://github.com/widelands/widelands this is a rough FP i fixed it.
(special thanks for it, I did not foresee it)
I added to the predicate `conditionsOutsideWrapper` the constraint `fcp instanceof ExprInVoidContext and`
3. example https://github.com/janisozaur/openrct2-nsf, 
As I understand it, it finds functions to handle emergencies.
this is an interesting situation requiring multiple call handling options.
I could limit the situation by tracking changes in the checked parameter (which is not typical for a security shell).
but to speed up the process, I added more coarse restrictions.
adding a call constraint to the `conditionsForWrapper` predicate.
</p>
</details>

**Comment by JarLob** [comment]
I have started a new run. We'll see.

**Comment by JarLob** [comment]
Hi @ihsinme,

Thanks for the submission! We have reviewed your report and validated your findings. After internally assessing the findings and the query we have determined this query doesn't meet the bar for scope and impact of the findings. Thus it is not eligible for a reward under the Bug Bounty program. We only award higher quality queries as part of the program with attribution on our Bug Bounty site and a payout between $500 and $3,000.
